### DISCUSSION TITLE
Can I use `useMutation` to get query fields as well?

### DISCUSSION THREAD

**Discussion by deadcoder0904** [dotansimha/graphql-code-generator#7102] [Q&A]
I have an `updateSettings` mutation:

```graphql
mutation UpdateSettings($userId: String!, $timezone: String) {
	updateSettings(userId: $userId, timezone: $timezone) {
		id
		timezone
	}
}
```

I want to update it but I also need `id` & when I use `useUpdateSettingsMutation()` then I can get `id` using:

```ts
const [{ data: settingsData }, updateSettings] = useUpdateSettingsMutation()

const id = settingsData?.updateSettings?.id
```

So should I use that or do I use query like `useSettingsQuery()`? What do people do generally?

I've never used codegen before & this is the 1st time I'm seeing mutation returning query fields as well so confused 🤔

**Comment by smolinari** [comment]
@deadcoder0904 

> I want to update it

Update what? 

And you can always get a return object/ selection of its fields (i.e. like the results after the mutation has been carried out), if your mutation resolver on the server returns the mutated object (in your example, `Settings`??). So, the normal process is, you should only run the mutation and if you send back the mutated object to the client (or selected fields), all you should be doing is refreshing the cache with it (or rather the client would do it for you). You should actually be updating your UI with the change directly. It's called optimistic responses. 

https://www.apollographql.com/docs/react/performance/optimistic-ui/

Edit: this only works with a single object mutation (i.e. with an id and in Apollo client. I'm assuming that is what you are using). If you are updating multiple objects, you also have to update  the cache yourself. Apollo won't do it for you.

Scott

**Reply by deadcoder0904** [reply]
> Update what?

Update Settings obviously.

I know Optimistic UI, have done that quite a lot but I went with using query fields.
